mlbench_core/optim/pytorch/fp_optimizers.py

```diff
@@ -0,0 +1,329 @@
+        fp32_params = torch.empty(
+            nelem,
+            dtype=torch.float32,
+            device=torch.device("cuda" if self.use_cuda else "cpu"),
+        )
+        fp16_params = torch.empty(
+            nelem,
+            dtype=torch.float16,
+            device=torch.device("cuda" if self.use_cuda else "cpu"),
+        )
+
+        pointer = 0
+        for p in self.fp16_model.parameters():
+            nelem = p.numel()
+            fp32_params[pointer : pointer + nelem].copy_(p.data.view(-1))
+            fp16_params[pointer : pointer + nelem].copy_(p.data.view(-1))
+            pointer += nelem
+
+        fp32_params = torch.nn.Parameter(fp32_params, requires_grad=True)
+        fp32_params.grad = torch.autograd.Variable(
+            fp32_params.data.new(*fp32_params.size())
+        )
+
+        fp16_params = torch.nn.Parameter(fp16_params, requires_grad=True)
+        fp16_params.grad = torch.autograd.Variable(
+            fp16_params.data.new(*fp16_params.size())
+        )
+
+        return fp16_params, fp32_params
+
+    @staticmethod
+    def fp16_to_fp32_flat_grad(fp32_params, fp16_model):
+        pointer = 0
+        for p in fp16_model.parameters():
+            nelem = p.numel()
+            fp32_params.grad.data[pointer : pointer + nelem].copy_(p.grad.data.view(-1))
+            pointer += nelem
+
+    @staticmethod
+    def fp16_to_fp16_flat_grad(fp16_params, fp16_model):
+        fp16_params.grad.data = torch.cat(
+            [p.grad.data.view(-1) for p in fp16_model.parameters()]
+        )
+
+    @staticmethod
+    def fp32_to_fp16_grads(fp16_model, fp32_params):
+        pointer = 0
+        for p in fp16_model.parameters():
+            nelem = p.numel()
+            p.data.view(-1).copy_(fp32_params.data[pointer : pointer + nelem])
+            pointer += nelem
+
+    def backward_loss(self, loss):
+        loss *= self.loss_scale
+        loss.backward()
+
+    def step(self, closure=None):
+        """
+        Performs one step of the optimizer.
+        Applies loss scaling, computes gradients in fp16, converts gradients to
+        fp32, inverts scaling and applies optional gradient norm clipping.
+        If gradients are finite, it applies update to fp32 master weights and
+        copies updated parameters to fp16 model for the next iteration. If
+        gradients are not finite, it skips the batch and adjusts scaling factor
+        for the next iteration.
+
+        Args:
+            closure (callable, optional): A closure that reevaluates the model and returns the loss.
+        """
+
+        scaling_factor = self.loss_scale
+
+        # Aggregate gradients
+        self.agg(self.fp16_model, self.agg_mode)
+        # Cast fp16 params to fp32 for optimizer
+        self.fp16_to_fp32_flat_grad(self.fp32_params, self.fp16_model)
+
+        if scaling_factor != 1.0:
+            self.fp32_params.grad.data /= scaling_factor
+        start = datetime.now()
+        torch.cuda.synchronize()
```

Comment:
@IamTao Here is the line where it blocks. At every call for `step()`, all processes stall for between 3 and 5 seconds here.
I have been looking into that for a few days now. Not sure how to solve it. Maybe it is due to GPUs not being connected through NVLink ?